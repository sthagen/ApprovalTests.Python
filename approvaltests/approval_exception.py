class ApprovalException(Exception):
    def __init__(self, value: str) -> None:
        self.value = value

    def __str__(self):
        return repr(self.value)


class FrameNotFound(ApprovalException):
    pass
