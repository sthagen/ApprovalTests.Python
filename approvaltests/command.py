import os


class Command(object):

    def __init__(self, cmd):
        self.command = cmd

    @staticmethod
    def executable(cmd):
        return os.path.isfile(cmd) and os.access(cmd, os.X_OK)

    def locate(self):
        path, name = os.path.split(self.command)
        if path and self.executable(self.command):
            return self.command
        if True: # use 'where' to find the executable so it finds python and python.exe
            pass

        for path in os.environ["PATH"].split(os.pathsep):
            path=path.strip('"')
            exe = os.path.join(path, self.command)
            if self.executable(exe):
                return exe

        return None