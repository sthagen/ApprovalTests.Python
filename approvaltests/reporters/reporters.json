[
  [
    "BeyondCompare4x64",
    "C:/Program Files/Beyond Compare 4/BCompare.exe"
  ],
  [
    "BeyondCompare4",
    "C:/Program Files (x86)/Beyond Compare 4/BCompare.exe"
  ],
  [
    "BeyondCompare4Mac",
    "/usr/local/bin/bcompare"
  ],
  [
    "BeyondCompare3",
    "C:/Program Files (x86)/Beyond Compare 3/BCompare.exe"
  ],
  [
    "AraxisMergeWin",
    "C:/Program Files/Araxis/Araxis Merge/Merge.exe"
  ],
  [
    "AraxisMergeMac",
    "/Applications/Araxis Merge.app/Contents/Utilities/compare"
  ],
  [
    "WinMerge",
    "C:/Program Files (x86)/WinMerge/WinMergeU.exe"
  ],
  [
    "kdiff3",
    "/usr/bin/kdiff3"
  ],
  [
    "meld",
    "meld"
  ],
  [
    "DiffMerge",
    "/Applications/DiffMerge.app/Contents/MacOS/DiffMerge"
  ],
  [
    "PyCharm",
    "/Applications/PyCharm CE.app/Contents/MacOS/pycharm",
    ["diff"]
  ]
]
