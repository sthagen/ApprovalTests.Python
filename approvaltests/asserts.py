from approvaltests import get_default_namer, verify_with_namer, write_to_temporary_file, get_reporter, \
    StackFrameNamer


class FilePathNamer(StackFrameNamer):
    def __init__(self, file_path, extension=None):
        StackFrameNamer.__init__(self, extension)
        self.file_path = file_path

    def get_approved_filename(self, basename=None):
        return self.file_path


def assert_against_file(actual, file_path, reporter=None):
    namer = FilePathNamer(file_path)
    verify_with_namer(actual, namer, reporter)


def assert_equal_with_reporter(expected, actual, reporter=None):
    if actual == expected:
        return

    name = get_default_namer().get_file_name()
    expected_file = write_to_temporary_file(expected, name + '.expected.')
    actual_file = write_to_temporary_file(actual, name + '.actual.')
    get_reporter(reporter).report(actual_file, expected_file)
    raise AssertionError('expected != actual\n  actual: "{}"\nexpected: "{}"'.format(actual, expected))
