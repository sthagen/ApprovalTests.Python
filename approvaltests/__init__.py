from .version import *
from .utils import *
from .approvals import *
from .asserts import *
from .combination_approvals import *

__author__ = "Chris Lucian; Llewellyn Falco; Jim Counts; Tim Preece"
__version__ = version_number
