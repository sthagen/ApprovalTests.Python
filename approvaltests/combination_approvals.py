from itertools import product

from approvaltests import verify, verify_with_namer, get_default_namer


def verify_all_combinations(
    function_under_test, input_arguments, formatter=None, reporter=None
):
    """Run func with all possible combinations of args and verify outputs against the recorded approval file.

    Args:
        function_under_test (function): function under test.
        input_arguments: list of values to test for each input argument.  For example, a function f(product, quantity)
            could be tested with the input_arguments [['water', 'cola'], [1, 4]], which would result in outputs for the
            following calls being recorded and verified: f('water', 1), f('water', 4), f('cola', 1), f('cola', 4).
        formatter (function): function for formatting the function inputs/outputs before they are recorded to an
            approval file for comparison.
        reporter (approvaltests.reporter.Reporter): an approval reporter.

    Raises:
        ApprovalException: if the results to not match the approved results.
    """
    namer = get_default_namer()
    verify_all_combinations_with_namer(
        function_under_test, input_arguments, namer, formatter, reporter
    )


def verify_all_combinations_with_namer(
    function_under_test, input_arguments, namer, formatter=None, reporter=None
):
    """Run func with all possible combinations of args and verify outputs against the recorded approval file.

    Args:
        function_under_test (function): function under test.
        input_arguments: list of values to test for each input argument.  For example, a function f(product, quantity)
            could be tested with the input_arguments [['water', 'cola'], [1, 4]], which would result in outputs for the
            following calls being recorded and verified: f('water', 1), f('water', 4), f('cola', 1), f('cola', 4).
        namer (approvaltests.Namer): A namer that defines the name of received and approved files.
        formatter (function): function for formatting the function inputs/outputs before they are recorded to an
            approval file for comparison.
        reporter (approvaltests.reporter.Reporter): an approval reporter.

    Raises:
        ApprovalException: if the results to not match the approved results.
    """
    if formatter is None:
        formatter = args_and_result_formatter
    approval_strings = []
    for args in product(*input_arguments):
        try:
            result = function_under_test(*args)
        except Exception as e:
            result = e
        approval_strings.append(formatter(args, result))
    verify_with_namer("".join(approval_strings), namer=namer, reporter=reporter)


def args_and_result_formatter(args, result):
    return "args: {} => {}\n".format(repr(args), repr(result))
