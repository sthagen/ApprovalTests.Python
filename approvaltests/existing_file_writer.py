import shutil

from approvaltests.core import Writer


class ExistingFileWriter(Writer):
    def __init__(self, file_name: str) -> None:
        self.file_name = file_name

    def write_received_file(self, received_file: str) -> str:
        shutil.copyfile(self.file_name, received_file)
        return received_file