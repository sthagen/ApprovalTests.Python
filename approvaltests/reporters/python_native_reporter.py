#!/usr/bin/env python

import os
import sys
from difflib import unified_diff

from approvaltests.reporters import get_command_text, Reporter
from approvaltests.utils import ensure_file_exists


class PythonNativeReporter(Reporter):
    """
    A blocking reporter that outputs diff straight
    to standard output
    """
    def report(self, received_path, approved_path):
        ensure_file_exists(approved_path)
        print(calculate_diff(received_path, approved_path))
        return True


def calculate_diff(file1, file2):
    with open(file1) as f1:
        with open(file2) as f2:
            diff = unified_diff(f2.readlines(),
                                f1.readlines(),
                                os.path.basename(file2),
                                os.path.basename(file1))
            diff_string = "\n".join(diff)
            if diff_string.strip():
                approve = get_command_text(file1, file2)
                approve_cmd = "\n\nto approve this result:\n\n" + approve + "\n"
            else:
                approve_cmd = ""
            return diff_string + approve_cmd


if __name__ == '__main__':
    fileA, fileB = sys.argv[1:3]
    print(calculate_diff(fileA, fileB))
