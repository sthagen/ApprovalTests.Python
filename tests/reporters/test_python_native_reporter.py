import os

from approvaltests import verify
from approvaltests.reporters import GenericDiffReporterFactory

from approvaltests.reporters.python_native_reporter import *

factory = GenericDiffReporterFactory()


def test_files_identical(tmpdir):
    file1 = os.path.join(str(tmpdir), "a.received.txt")
    file2 = os.path.join(str(tmpdir), "b.approved.txt")
    identical_contents = "abc"
    with open(file1, "w") as f1:
        f1.write(identical_contents)
    with open(file2, "w") as f2:
        f2.write(identical_contents)
    verify(calculate_diff(file1, file2), reporter=factory.get("PythonNative"))


def test_files_differ(tmpdir):
    file1 = os.path.join(str(tmpdir), "a.received.txt")
    file2 = os.path.join(str(tmpdir), "b.approved.txt")
    with open(file1, "w") as f1:
        f1.write("abc")
    with open(file2, "w") as f2:
        f2.write("def")
    diff = calculate_diff(file1, file2)
    diff = diff.replace(str(tmpdir), "tmpdir")  # use scrubber in future
    diff = diff.replace('\\', '/')

    verify(diff, reporter=factory.get("PythonNative"))

