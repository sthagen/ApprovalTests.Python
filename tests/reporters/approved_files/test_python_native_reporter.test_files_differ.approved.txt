--- b.approved.txt

+++ a.received.txt

@@ -1 +1 @@

-def
+abc

to approve this result:

mv -f tmpdir/a.received.txt tmpdir/b.approved.txt
